### ISSUE TITLE
Encode strings when writing to shared memory

### ISSUE THREAD

**Issue by bschmalhofer** [RotherOSS/otobo#1082]
`Kernel::System::Log::Log()` is using calling `shmwrite` for writing the last log entries to shared memory. The entries can be retrieved with `Kernel::System::Log::GetLog()`. GetLog assumes that the read in data is UTF-8 encoded. The method decodes the binary data into proper Perl strings. Howevert Log() does not UTD-8 encode the data written to shared memory. Using Perl up the 5.32 this coincidentally worked mostly fine. No problems the strings are already UTF-8 encoded internally, or when the string is strictly ascii. There might have been unnoticed problems for characters in the Latin 1 - Supplement. See https://metacpan.org/dist/perl/view/pod/perlunicode.pod#The-%22Unicode-Bug%22.

The situation is different when running under Perl 5.34. In the changelog https://perldoc.perl.org/5.34.0/perldelta  there is the entry:

> semop() now downgrades the OPSTRING parameter, msgsnd() now downgrades the MSG parameter and shmwrite now downgrades the STRING parameter to treat them as bytes. Previously they would be left upgraded, providing a corrupted structure to the underlying function call.

This means the shmwrite implicitly calls utf8::downgrade() on the argument. This fails for strings with wide character und in consquence `Kernel::System::Log()` throws an exception.

> Error: Wide character in shmwrite at /opt/otobo/Kernel/System/Log.pm line 276.

The solution is simple: 

- [x] UTF-8 encode the string written in Kernel::System::Log::Log() to shared memory

See also #1081

**Comment by bschmalhofer** [comment]
Looking at the usage of EncodeOutput in OTOBO I noticed at least one fishy instance.
```
        if ( $ConfigObject->Get('Frontend::AvatarEngine') eq 'Gravatar' && $Self->{UserEmail} ) {
            my $DefaultIcon = $ConfigObject->Get('Frontend::Gravatar::DefaultImage') || 'mp';
            $Kernel::OM->Get('Kernel::System::Encode')->EncodeOutput( \$Self->{UserEmail} );
            $Param{Avatar} = '//www.gravatar.com/avatar/' . md5_hex( lc $Self->{UserEmail} ) . '?s=100&d=' . $DefaultIcon;
        }
```
In that example `$Self->{UserEmail} ` is binary. Calling `lc` on binary data is fishy. That most likely never caused an error, as emails are usually ASCII only.

**Comment by bschmalhofer** [comment]
Added the EncodeOutput() before shmwrite. Test suite showed the same errors as HEAD of rel-10_1. Merged the PR. Closing the issue.

**Comment by bschmalhofer** [comment]
The issue #1082 fixes the shmwrite problem in the OTOBO 10.1.x series. But Perl 5.34 can also be used for running the 10.0.x series. This means that _Kernel/System/Log.pm_ must be adapted in rel-10_0 as well. 

See also: https://otobo.de/de/forums/topic/nach-migration-kommt-bei-der-anmeldung-nur-internal-server-error/#post-13561
